### DISCUSSION TITLE
Convert column to list, but only it's not already a list?

### DISCUSSION THREAD

**Discussion by gotdan** [duckdb/duckdb#13347] [General]
Is there a function to convert a column that's not already a list into a list, but leave existing lists alone?

Essentially, I'd like to do something like `CREATE MACRO as_list(a) AS CASE WHEN typeof(a)[-2:-1] = '[]' THEN a ELSE list(a) END;`. Unfortunately, this doesn't work and I get an error `Cannot mix values ... in CASE expression`. 

Is there a built in function I'm missing or a way to work around the error with the macro?

Thanks!

**Comment by pkoppstein** [comment]
The values in a column are constrained by the declared type of the column, so using the standard types (e.g. INTEGER), the situation you describe couldn't occur .... but the JSON type provides a possible workaround:
```
SELECT CASE WHEN typeof(i)[-2:-1] = '[]' THEN CAST(i AS JSON) ELSE cast([i] AS JSON) END from t;
```
This technique of course assumes you don't mind having a JSON column.

**Comment by mjs1013** [comment]
Did you end up finding a solution to this issue? In my project, I am trying to apply a macro to an arbitrary column specfied via a column name passed in as a macro parameter.. So if the selected column is of type varchar[], I'd like to flatten the varchar[] and then perform my logic as if it was a typical varchar column. However, I'm running into the same case error when attempting to flatten if typeof like [], else select the column and proceed.

**Comment by soerenwolfers** [comment]
@mjs1013  @gotdan 
Everything is possible with https://duckdb.org/docs/data/json/sql_to_and_from_json.html and the more recent addition https://duckdb.org/docs/sql/functions/utility.html#queryquery_string_literal  but I wouldn't unconditionally recommend doing this. 

For example: 

```sql
CREATE OR REPLACE MACRO to_list(a) AS (SELECT * FROM query('SELECT ' || CASE WHEN typeof(a)[-2:-1] = '[]' THEN a::VARCHAR ELSE [a]::VARCHAR END));
SELECT to_list([1]), to_list(1)
```
```
┌─────────────────────────────┬────────────┐
│ to_list(main.list_value(1)) │ to_list(1) │
│           int32[]           │  int32[]   │
├─────────────────────────────┼────────────┤
│ [1]                         │ [1]        │
└─────────────────────────────┴────────────┘
```

**Reply by mjs1013** [reply]
This is exactly what I was looking for. Thank you for the example!

**Comment by soerenwolfers** [comment]
There is now a "proper" solution for this via https://github.com/duckdb/duckdb/pull/17209
